### PR TITLE
[feat] Add HunyuanVideo 1.5 T2V training support

### PR DESCRIPTION
**Pull request by KyleNeverGivesUp**
## Purpose

Adds HunyuanVideo 1.5 to FastVideo's modular training stack, so teams already training Wan/Cosmos here can fine-tune HY1.5 with the same YAML and preprocessed-parquet workflow, and so HY1.5 becomes eligible for the framework's distillation and RL methods.

Related: #1661 (an inference-side ByT5 sizing bug found while running validation here; not fixed in this PR, see Limitations below).

> **Note:** depends on #1663 (@klhhhhh) for the HY1.5 data-side pipeline: the dual text-embedding parquet schema, its collate path, and the preprocess script. The tests in this PR use synthetic batches and pass standalone; only the end-to-end overfit run needs that PR's output.

## Changes

`Hunyuan15Model` subclasses `WanModel` and overrides only what differs:

- **`prepare_batch`**: dual text embeddings (Qwen 3584 plus ByT5 1472) with a zero-token fallback for parquet that predates the ByT5 fields, padding trimmed to the batch's longest valid length, and `"hunyuan15"` VAE scaling (the parquet stores raw `latent_dist.mode()` outputs). Padding is trimmed rather than masked because the HY1.5 DiT has no text mask input and its token refiner mean-pools over every token; with `train_batch_size: 1` this matches inference exactly.
- **`predict_noise` and `_build_distill_input_kwargs`**: the HY1.5 forward contract. `(B, C, T, H, W)` layout with conditioning channels packed to 65 (32 latent, 1 mask, 32 conditioning latent, the last two all zero for T2V, mirroring `Hy15ImageEncodingStage` and `DenoisingStage`), `encoder_hidden_states` as `[qwen, byt5]`, an all-zero SigLIP-shaped placeholder to select the T2V branch, and the discrete 0..1000 timestep. `encoder_attention_mask`, `return_dict`, and `timestep_r` are deliberately absent, since HY1.5's forward rejects or crashes on them.
- **`ensure_negative_conditioning`**: encodes `"."` through the same Qwen chat-template path the preprocess script uses. The shared `encode_negative_prompt` helper does not fit, because HY1.5's preprocess returns a chat message list, its postprocess takes the attention mask and returns a tuple, and it reads `hidden_states[-3]`.
- **`init_preprocessors`**: `Qwen2_5_VLArchConfig` leaves `text_len` at 0, which would size the dataloader's text padding to zero tokens.

Supporting changes:

- `normalize_dit_input` gains a `"hunyuan15"` branch (`latents * scaling_factor`).
- Single-GPU overfit config (LoRA) and 8-GPU full-parameter fine-tuning config.
- CPU contract tests plus CUDA-gated load and single-step tests, both covering the zero-token ByT5 case.
- Training runbook at `docs/training/hunyuan15_overfit.md`.
- `fastvideo/tests/train/methods/grad_norm_regression.py` gains `double_blocks` in `_BLOCK_LIST_ATTRS`. That list already exists to cover per-family naming; the entry is additive and changes nothing for models using `blocks` or `transformer_blocks`.

## Test Plan

```bash
# Unit and CUDA-gated tests (A100 80GB)
python -m pytest fastvideo/tests/train/ -v -k hunyuan15

# Data pipeline
CUDA_VISIBLE_DEVICES=0 python fastvideo/pipelines/preprocess/preprocess_hunyuan15_overfit.py

# Single-clip overfit
export WANDB_MODE=offline
bash examples/train/run.sh examples/train/configs/overfit_hunyuan15_t2v.yaml

# Lint
pre-commit run --all-files
```

## Test Results

<details>
<summary>pytest, 12 passed on A100 80GB</summary>

```
fastvideo/tests/train/methods/test_hunyuan15_finetune.py::test_hunyuan15_finetune_single_train_step[with_byt5] PASSED
fastvideo/tests/train/methods/test_hunyuan15_finetune.py::test_hunyuan15_finetune_single_train_step[empty_byt5] PASSED
fastvideo/tests/train/models/test_hunyuan15_smoke.py::test_kwargs_match_transformer_signature PASSED
fastvideo/tests/train/models/test_hunyuan15_smoke.py::test_forbidden_kwarg_absent[encoder_attention_mask] PASSED
fastvideo/tests/train/models/test_hunyuan15_smoke.py::test_forbidden_kwarg_absent[return_dict] PASSED
fastvideo/tests/train/models/test_hunyuan15_smoke.py::test_forbidden_kwarg_absent[timestep_r] PASSED
fastvideo/tests/train/models/test_hunyuan15_smoke.py::test_dual_text_embeddings_are_a_two_element_list PASSED
fastvideo/tests/train/models/test_hunyuan15_smoke.py::test_zero_token_byt5_is_preserved PASSED
fastvideo/tests/train/models/test_hunyuan15_smoke.py::test_t2v_image_placeholder_is_all_zero PASSED
fastvideo/tests/train/models/test_hunyuan15_smoke.py::test_hidden_states_are_packed_to_65_channels PASSED
fastvideo/tests/train/models/test_load_hunyuan15.py::test_hunyuan15_model_loads_and_forwards PASSED
fastvideo/tests/train/models/test_load_hunyuan15.py::test_hunyuan15_forwards_with_empty_byt5 PASSED

12 passed
```

The 8 CPU tests check the assembled forward kwargs against `HunyuanVideo15Transformer3DModel.forward`'s real signature via `inspect.signature`, so signature drift fails loudly without loading weights. The 4 CUDA tests load the real 8.33B checkpoint: two run a forward pass, two run a full training step and assert block 0's gradients are finite and non-zero.

</details>

<details>
<summary>Preprocess, parquet contract</summary>

```
[Qwen 1/1] (29, 3584)          # caption tokens x Qwen hidden size
[ByT5 1/1] (0, 1472)           # zero-length: no quoted glyph text in the caption
[VAE  1/1] (32, 21, 30, 52)    # 32 channels, 21 latent frames, 480/16 x 832/16
Wrote 1 records to data/hunyuan15_overfit_preprocessed/data_00000.parquet
```

</details>

<details>
<summary>Single-clip overfit, loss</summary>

A100 80GB, LoRA rank 64 with AdamW, 180 steps on one clip. Flow-matching loss is noisy step to step because each step samples a random timestep, so these are segment means:

```
segment 1/6: 0.1551
segment 2/6: 0.1673
segment 3/6: 0.0626
segment 4/6: 0.1443
segment 5/6: 0.1111
segment 6/6: 0.0746

first quarter: 0.1507
last  quarter: 0.0805     (-47%)
```

The same code path also runs end to end on aarch64 with Blackwell (DGX Spark, GB10, CUDA 13.0) in full-parameter mode.

</details>

## Limitations

- T2V only; I2V is out of scope here.
- Full-parameter 8B training needs roughly 80GB before activations, so it does not fit one 80GB card. The overfit config uses LoRA and the 8-GPU config uses FSDP. The runbook has the memory breakdown.
- CFG dropout semantics are not aligned yet: the data pipeline zeroes the embedding values while inference encodes `"."`. This only matters at `training_cfg_rate > 0`, and both configs set it to 0. `ensure_negative_conditioning` already implements the inference semantics, so aligning the collate side later is self-contained.
- The overfit config leaves validation disabled because of #1661, where the training-side validation pipeline overwrites the loaded text-encoder config and mis-sizes the zero-length ByT5 placeholder. Standalone HY1.5 inference is unaffected.

## Checklist

- [x] I ran `pre-commit run --all-files` and fixed all issues
- [x] I added or updated tests for my changes
- [x] I updated documentation if needed
- [x] I considered GPU memory impact of my changes

**For model/pipeline changes, also check:**
- [ ] I verified SSIM regression tests pass. Not applicable: this PR adds training only and does not touch any inference path.
- [ ] I updated the support matrix if adding a new model. HY1.5 inference support already exists; this PR adds training.

### CODE REVIEW

File: fastvideo/train/models/hunyuan15/hunyuan15.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+        default of 0, which would size the parquet dataloader's text
+        padding to zero tokens.  The real padded length is the Qwen
+        tokenizer max length minus the cropped template tokens
+        (1108 - 108 = 1000), matching the preprocess output.
+        """
+        pipeline_config = training_config.pipeline_config
+        assert pipeline_config is not None
+        text_len = int(pipeline_config.text_encoder_max_lengths[0] - pipeline_config.text_encoder_crop_start)
+        pipeline_config.text_encoder_configs[0].arch_config.text_len = (text_len)
+        super().init_preprocessors(training_config)
```
[the remark this answers is not in the document]
Comment (KyleNeverGivesUp) [reply fastvideo/train/models/hunyuan15/hunyuan15.py:99]: Depends on #1663, which adds the text_embedding_2 fields to the schema and collate path. Until it lands, prepare_batch takes the zero-token fallback on main, which is why the tests here use synthetic batches.

File: fastvideo/train/models/hunyuan15/hunyuan15.py [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+from fastvideo.training.training_utils import (
+    normalize_dit_input, )
+
+if TYPE_CHECKING:
+    from fastvideo.train.utils.lora import LoraConfig
+    from fastvideo.train.utils.training_config import (
+        TrainingConfig, )
+
+
+class Hunyuan15Model(WanModel):
```
[the remark this answers is not in the document]
Comment (KyleNeverGivesUp) [reply fastvideo/train/models/hunyuan15/hunyuan15.py:40]: Fixed in 927938f with a scaling-factor decode override.

File: fastvideo/train/models/hunyuan15/hunyuan15.py [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+        raw_batch: dict[str, Any],
+        *,
+        generator: torch.Generator,
+        latents_source: Literal["data", "zeros"] = "data",
+    ) -> TrainingBatch:
+        """Same flow as Wan, with three HY1.5 differences: dual text
+        embeddings (Qwen + ByT5), trim-to-longest padding removal (the
+        DiT has no text-mask input) and "hunyuan15" VAE scaling."""
+        if self._requires_negative_conditioning:
+            self.ensure_negative_conditioning()
```
[the remark this answers is not in the document]
Comment (KyleNeverGivesUp) [reply fastvideo/train/models/hunyuan15/hunyuan15.py:146]: Fixed in ad1ca9b5.

**Comment by SolitaryThinker** [comment]
Kyle — pushed a commit to your branch rather than sending a list back. Revert anything you disagree with.

**The main one: the ByT5 fallback was silent.** `hunyuan15.py:146` absorbed missing `text_embedding_2` into a `[B, 0, 1472]` zero-token tensor and logged nothing. That's indistinguishable from healthy training — finite grads, falling loss, full speed — and `docs/training/hunyuan15_overfit.md:81-84` tells the reader a `(0, 1472)` shape is normal and expected, so the broken state and the working state look identical.

It's also reachable for a reason unrelated to legacy data. `collate_rows_from_parquet_schema` derives its tensor fields from the parquet schema, so if the schema doesn't declare the `text_embedding_2` columns the field is dropped before `prepare_batch` sees it — even when the row carries it. Measured on this branch alone with a row that does carry `text_embedding_2_bytes`:

```
schema has text_embedding_2_bytes: False
raw_batch.get('text_embedding_2') -> NoneType
```

Those columns come from #1663. I merged it locally and re-measured: `shape=(1, 1000, 1472)`, `mask_sum=11`, trimmed to `[1, 11, 1472]` — correct. So this isn't a bug in your code, it's that nothing detects when the contract isn't met, and merge order isn't enforced anywhere. If this lands first, every HY1.5 run trains with glyph conditioning off and looks fine. Added a warning on that branch, once per model rather than per step.

**Also replaced the two `1472` literals with the config that owns the value.** `HunyuanVideo15ArchConfig.text_embed_2_dim` is what constructs the `nn.LayerNorm` the tensor has to satisfy, so it's the authority — reachable as `tc.pipeline_config.dit_config.arch_config.text_embed_2_dim`, and the base class already reads `dit_config.patch_size` from the same place. The comment said the width couldn't be derived from config because `T5ArchConfig` defaults to 512; that's true of the encoder config but it isn't where this width comes from. Verified the config returns 1472, so it's behavior-preserving.

**And `check_grad_norm_regression` is now called** in `test_hunyuan15_finetune.py`. It was imported and never used — the only one of the six per-method tests that didn't call it (ruff flags F401; test files are excluded from the lint lane, so CI stayed green). It skips cleanly on a device with no seeded reference, so it costs nothing now and guards once one exists.

Verified: ruff clean on the changed files, CPU tests 10 passed.

Credit where it's due — the subclass boundary is right, overriding only the four methods that actually differ from Wan, and the one shared-file change is a new branch in a dispatch that already carries per-family cases. I checked the forward contract against the inference path rather than taking it on faith: `Hy15ImageEncodingStage` emits the same all-zero `[1, 729, 1152]` placeholder, denoising concatenates the same `32 + 1 + 32 = 65` channels, and the transformer's `is_t2v` check confirms the all-zero content selects T2V. The token refiner really is called without a mask, so trimming rather than masking is correct. `init_preprocessors` is necessary and the `1108 - 108 = 1000` arithmetic matches the config. The runbook is good — the memory table and the honest troubleshooting entry for the ByT5 sizing issue in particular.

One note, not a request: both configs leave validation commented out, but the templates use `every_steps` with a `% == 0` gate and validation runs once before the loop at step 0 — whoever uncomments them gets a full sampling pass before training starts.

Not verified here: the end-to-end overfit (needs #1663's preprocess script), the 8-GPU FSDP config, and `ensure_negative_conditioning`, which no shipped config reaches.

**Comment by KyleNeverGivesUp** [comment]
Thanks, keeping all four. The width one corrects me that the authority is the LayerNorm text_embed_2_dim builds, not the encoder config I was reading.

**Comment by KyleNeverGivesUp** [comment]
@SolitaryThinker this has been ready since your commits landed. All
review threads are resolved and pre-commit is green, so it just needs an
approve.

**Comment by KyleNeverGivesUp** [comment]
Rebased onto current main, zero commits behind. The earlier failure was Build
#4732 from Aug 14, on a base that was 41 commits stale. The rebase was clean:
this PR is almost entirely new files, with three lines touching existing code.

Ran this PR's tests on a DGX Spark. `test_load_hunyuan15.py` and
`test_hunyuan15_smoke.py`: 10 of 10 pass.

`test_hunyuan15_finetune_single_train_step` reports SKIPPED, but the skip comes
from `grad_norm_regression.py:168`, not from the CUDA guard at the top of the
test:

```
SKIPPED [1] fastvideo/tests/train/methods/grad_norm_regression.py:168: no
grad-norm reference for test_hunyuan15_finetune on 'NVIDIA GB10'
(device_key=None); run with FASTVIDEO_GRADNORM_UPDATE=1 to seed it
```

So the step itself runs, including the assertion that layer-0 grads are
non-zero, and only the final device-keyed comparison is skipped. 15.8s for a
single case.

GB10 is not in `_DEVICE_MAPPINGS`, so there is no key to look up. Happy to seed
a reference for it separately if that is useful; it would turn this check into a
real assertion on that hardware rather than a skip.
